README.md
# yandex-rss-stats

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-05-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-05-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/yandex-rss-stats/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/yandex-rss-stats/compare/0.1.0...0.1.1


test/yandex_rss_stats/core_test.clj
(ns yandex-rss-stats.core-test
  (:require [clojure.test :refer :all]
            [yandex-rss-stats.core :refer :all]
            [org.httpkit.server    :refer [with-channel send!]]
            [org.httpkit.client :as http]))

(defn with-server-started [f]
  (letfn [(echo-async-handler [{:keys [async-channel body]}]
            (send! async-channel {:status 200
                                  :body   body})
            {:body async-channel})]
    (with-redefs [yandex-rss-stats.controller/handler echo-async-handler]
      (start-server)
      (try
        (f)
        (finally
          (stop-server))))))

;; test that server supports async handlers
(deftest a-test
  (testing "/search"
    (is (= "ping"
           (:body @(http/post "http://localhost:8080/foo/bar" {:body "ping"
                                                               :as   :text}))))))

(use-fixtures :each with-server-started)


test/yandex_rss_stats/stats_test.clj
(ns yandex-rss-stats.stats-test
  (:require [clojure.test :refer :all]
            [yandex-rss-stats.stats :refer :all]))

(deftest make-stats-test
  (are [input output] (= (make-stats input) output)
       ;; empty input give empty
       {} {}

       ;; query with no results give zero
       {"foo" []} {}

       ;; 2nd domain name is extracted
       {"foo" ["http://www.example.com/path/and?some=query&params"]} {"example.com" 1}
       {"foo" ["https://www.example.com"]} {"example.com" 1}
       {"foo" ["http://x.y.z.www.example.com"]} {"example.com" 1}

       ;; check counts
       {"foo" ["http://www.example.com/one" "http://example.com/two"]} {"example.com" 2}

       ;; check duplicates
       {"foo" ["http://example.com/same" "http://example.com/same"]} {"example.com" 1}

       ;; check multiple queries
       {"foo" ["http://example.com/one"
               "http://example.com/two"]
        "bar" ["http://example.com/three"]}
       {"example.com" 3}))


test/yandex_rss_stats/controller_test.clj
(ns yandex-rss-stats.controller-test
  (:require [clojure.test :refer :all]
            [yandex-rss-stats.controller :refer :all]
            [org.httpkit.client :as http]
            [ring.mock.request :as mock]
            [cheshire.core :refer [parse-string]]))

(deftest handler-test

  (testing "404"
    (let [result (handler (mock/request :get "/foo"))]
      (is (= (:status result) 404))
      (is (= (:body result) "Use /search end point"))))

  (testing "vanilla mock"
    (let [mock-channel      'mock-channel
          send!-calls       (atom [])
          blog-search-calls (atom [])
          make-stats-calls  (atom [])
          req               (assoc (mock/request :get "/search?query=foo&query=bar")
                              :async-channel mock-channel)]

      ;; replace send! and blog-search functions with mocked implementation
      (with-redefs [org.httpkit.server/send!                (fn [& args]
                                                              (swap! send!-calls conj args))
                    yandex-rss-stats.yandex-api/blog-search (fn [& args]
                                                              (swap! blog-search-calls conj args))
                    yandex-rss-stats.stats/make-stats       (fn [& args]
                                                              (swap! make-stats-calls conj args)
                                                              {:stats "stub stats"})]

        ;; invoke the handler
        (let [ret (handler req)]

          ;; check that channel response was returned
          (is (= {:status  200
                  :headers {}
                  :body mock-channel}
                 ret)))

        ;; check blog-search invocations
        (let [[[arg-1-1 arg-1-2] [arg-2-1 arg-2-2]] @blog-search-calls]

          ;; check how many times it was invoked
          (is (= (count @blog-search-calls) 2))

          ;; check how many arguments were passed
          (for [args @blog-search-calls]
            (is (= (count args) 2)))

          ;; first argument should come from query params
          (is (= arg-1-1 "foo"))
          (is (= arg-2-1 "bar"))

          ;; check that send! isn't yet invoked (handler should wait for the client to respond)
          (is (= [] @send!-calls))

          ;; at this point in time, we emulate that the second client is ready
          (arg-2-2 true ["link B1", "link B2"])

          ;; send! should not yet be invoked
          (is (= [] @send!-calls))

          ;; now the first client is ready
          (arg-1-2 true ["link A1", "link A2"]))

        ;; TODO wait for condition using clj-async-test
        (Thread/sleep 1000)

        ;; check send! invocations
        (let [[[arg1 arg2]] @send!-calls]

          ;; check how many times it was invoked
          (is (= (count @send!-calls) 1))

          ;; check how many arguments were passed
          (is (= (count (first @send!-calls)) 2))

          ;; check the first argument
          (is (= arg1 mock-channel))

          ;; check the second argument
          (is (= (-> arg2
                     :body
                     parse-string)
                 {"stats" "stub stats"}))
          (is (= {:status 200
                  :headers {"Content-Type" "application/json"}}
                 (dissoc arg2 :body))))

        ;; check make-stats invocation
        (let [[[arg]] @make-stats-calls]
          (is (= {"foo" ["link A1" "link A2"]
                  "bar" ["link B1" "link B2"]}
                 arg))))))

  ;; FIXME get rid of boilerplate code
  (testing "single query"
    (let [mock-channel      'mock-channel
          send!-calls       (atom [])
          blog-search-calls (atom [])
          req               (assoc (mock/request :get "/search?query=foo")
                              :async-channel mock-channel)]

      ;; replace send! and blog-search functions with mocked implementation
      (with-redefs [org.httpkit.server/send!                (fn [& args]
                                                              (swap! send!-calls conj args))
                    yandex-rss-stats.yandex-api/blog-search (fn [& args]
                                                              (swap! blog-search-calls conj args))
                    yandex-rss-stats.stats/make-stats (fn [& args] "stub")]

        ;; invoke the handler
        (handler req)

        ;; check that blog-saarch wasn't invoked for each of \f \o \o letters
        (let [[[arg1 arg2] :as args] @blog-search-calls]
          ;; check how many times it was invoked
          (is (= (count args) 1))
          (is (= arg1 "foo"))))))

  (testing "client failure"
    (let [mock-channel 'mock-channel
          send!-calls  (atom [])
          blog-search-calls (atom {})
          req          (assoc (mock/request :get "/search?query=foo&query=bar")
                         :async-channel mock-channel)]
      (with-redefs [org.httpkit.server/send!                (fn [& args]
                                                              (swap! send!-calls conj args))
                    yandex-rss-stats.yandex-api/blog-search (fn [query callback]
                                                              (swap! blog-search-calls assoc query callback))
                    yandex-rss-stats.stats/make-stats       (fn [& args] (is false "make-stats should not be invoked"))]

        ;; invoke the handler
        (handler req)

        ;; wait until client is invoked
        ;; TODO use clj-async-test
        (Thread/sleep 100)
        (is (= (count @blog-search-calls) 2))

        ;; mock responses from client
        (let [{:strs [foo bar]}  @blog-search-calls]
          ;; the foo call will succeed
          (foo true ["linke1", "linke2"])
          ;; the bar call will fail
          (bar false "I'm failed"))

        ;; give it time to think and invoke send! function
        (Thread/sleep 100)
        (is (= (count @send!-calls) 1))

        (let [[[arg1 arg2]] @send!-calls]
          (is (= arg1 mock-channel))
          (is (= arg2 {:status 500
                       :body   "Query bar failed"})))))))


src/yandex_rss_stats/core.clj
(ns yandex-rss-stats.core
  (:require [clojure.tools.logging :as log]
            [org.httpkit.server    :refer [run-server]]
            [compojure.core        :refer [defroutes GET]]
            [compojure.route       :as route])
  (:require [yandex-rss-stats.controller :refer [handler]]))

(defonce server (atom nil))

(add-watch server :log-server-state (fn [key ref old new]
                                      (if-let [port  (->> new
                                                          meta
                                                          :local-port)]
                                        (log/info "Server started on port" port)
                                        (log/info "Server stopped"))))

(defn start-server []
  (swap! server (fn [current]
                  (assert (nil? current) "Already started")
                  (run-server handler {:port 8080}))))

(defn stop-server []
  (swap! server (fn [stop-fn]
                  (assert stop-fn)
                  (stop-fn)
                  nil)))

(defn -main []
  (log/info "Application started")
  (start-server))


src/yandex_rss_stats/yandex_api.clj
(ns yandex-rss-stats.yandex-api
  (:require [clojure.tools.logging :as log]
            [org.httpkit.client :as http]
            [clj-xpath.core   :refer :all]))

(def YANDEX_API_URL "https://yandex.ru/blogs/rss/search")

(defn blog-search [query callback]
  (letfn [(on-response [{:keys [status body error] :as res}]
            ;; FIXME try-catch inside letfn looks ugly, other options?
            ;; TODO check `error`
            (try
              (if (= 200 status)
                (let [result ($x:text* "/rss/channel/item/link" body)]
                  (callback true result))
                (callback false "Unexpected status code"))
              (catch Throwable e
                (callback false e))))]
    ;; TODO limit number or parallel connections
    (http/get YANDEX_API_URL {:query-params {"text" query}} on-response)))




src/yandex_rss_stats/controller.clj
(ns yandex-rss-stats.controller
  ;; TODO rename to handler
  (:require [clojure.tools.logging :as log]
            [org.httpkit.server    :refer [with-channel send!]]
            [compojure.core        :refer [routes GET]]
            [compojure.route       :as route]
            [cheshire.core         :refer [generate-string]]
            [ring.middleware.params :refer [wrap-params]]
            [clojure.core.async         :refer [chan go >!! <! alt!] :as async])
  (:require [yandex-rss-stats.yandex-api :refer [blog-search]]
            [yandex-rss-stats.stats :refer [make-stats]]))

(defn- search
  "Controller function for /search endpoint"
  [{{:strs [query]} :query-params, :as ring-request}]
  (with-channel ring-request channel
    (let [query   (if (string? query) ;; treat single query as a singleton array
                    [query]           ;; TODO put it shorter: (flatten [query])
                    query)
          n       (count query)
          results (chan n)
          failure (chan 1)]

      ;; call blog search
      (doseq [query-elem query]
        (blog-search query-elem (fn [ok? links] ;; TODO can we use multimethods here?
                                  (if ok?
                                    (do
                                      (log/info "Completed search for" query-elem)
                                      (>!! results [query-elem links]))
                                    (do
                                      (log/error "Failed search for" query-elem)
                                      (>!! failure query-elem))))))

      ;; get search results and make the response
      (let [aggregated-result (->> results
                                   ;; close channel after all elements received
                                   (async/take n)
                                   ;; squash n elements into one collection
                                   (async/into {}))
            ;; put var into closure in order to force current thread bindings to take effect inside async code
            ;; (needed for unit tests)
            send! send!]
        (go (alt! aggregated-result ([result] (let [body (-> result              ;; get collection of results
                                                             make-stats                         ;; calculate stats
                                                             (generate-string {:pretty true}))] ;; serialize
                                                (send! channel {:status  200 ;; TODO remove default status
                                                                ;; TODO use middleware to add content type
                                                                :headers {"Content-Type" "application/json"}
                                                                :body    body})))
                  failure           ([elem] (send! channel {:status 500
                                                            :body   (str "Query " elem " failed")}))))))))

(def ^:private unwrapped-routes
  (routes
    (GET "/search" [:as req]
      (search req))
    (route/not-found "Use /search end point")))

(def handler (wrap-params unwrapped-routes))
